### DISCUSSION TITLE
De-duplication of minor roads

### DISCUSSION THREAD

**Discussion by davecraig** [onthegomap/planetiler#1082] [Q&A]
Hi,

Firstly, apologies if this has been discussed before or the answer is obvious - I've searched and pondered and come up with nothing.

We're generating a protomaps file to be served up to our app and it's used in two ways:

1. The tiles are displayed in the graphical UI with maplibre native on Android
2. The tiles at the maximum zoom level are decoded to GeoJSON to drive an audio description for the visually impaired. This includes describing the immediate surroundings including roads, road intersections and POIs.

For the second item we've been hoping to just use stock `planetiler` and using `--simplify-tolerance-at-max-zoom=0` we can generate tiles in which we can find all of the intersections. The problem we have is that we end up with duplication of roads from the `transportation` and `transportation_name` layers and I haven't found a way to successfully de-duplicate them.

My first question is why are the roads duplicated? Is it that it's just what the openmaptiles spec requires? I assume that for drawing graphical maps this doesn't matter?

Given that they are duplicated, how do I de-duplicate them? I think the main underlying issue is that the transportation layer merges many paths/minor roads etc. into `MultiStrings` which is great, except the `id` for those Features is just for one of the `LineStrings` and so can't be used to de-duplicate lines from the `transportation` layer with lines from the `transportation_name` layer.

If I alter `Transportation.postProcess` to add the id() as a tag on every item when at the max zoom level:

```
      if (zoom == config.maxzoom()) {
        item.tags().put(OSM_ID_MERGE_TAG, item.id());
      }
```
it seems to do what I want. I get a feature with a `LineString` for each OSM id and so obviously the file size goes up (by ~2% on my test map of Scotland), but given what I want to do, is this a reasonable approach? 

Many Thanks,

Dave Craig.
https://github.com/Scottish-Tech-Army/Soundscape-Android

**Comment by msbarry** [comment]
I'm not exactly sure why transportation and transportation_name are separate layers in openmaptiles. Typically styles render transportation as lines and transportation_name as text or symbols. You could either add names to transportation, or add ids to transportation_name (and set a smaller min size) to avoid combining them. 

Another option if you're not too tied to openmaptiles would be switching to protomaps which just has one "roads" layer with names: https://github.com/protomaps/basemaps/tree/main/tiles

**Reply by davecraig** [reply]
Thanks for the swift answers.  I'll stick with the openmaptiles for now and see how it goes. I like OSM Bright and don't want to get too deep into map styling (for now!).

Thanks
Dave

**Comment by zerda-ocm** [comment]
> My first question is why are the roads duplicated?

My educated guess is: to fit more labels. Imagine a "Main Street", that has different attributes every few meters, like speed limits, surface, lanes, etc. A renderer like maplibre treats each section individually and would try to fit the street name on each tiny section, whereas on transportation_name has only one attribute, and sections of the road can easily combined into one single line and therefore make it easier to find a big enough space for the text.

**Reply by davecraig** [reply]
Agreed, it's making a bit more sense to me now. The duplication of all of the transportation lines seems like a lot, but probably pales into insignificance next to the number of polygons for buildings.

My initial change which was adding of the OSM ids to the `transportation` layer initially looked like it worked, but then I found that of course the `transportation_name` layer is combining multiple OSM ids into a single entity which resulted in not all of the lines within the `transportation` layer having a name mapped to their OSM id. That's the whole point of that layer, and changing it would impact the graphical rendering. The solution for me is therefore to add the name to every line in the `transportation` layer. This should be okay as:

1. Maplibre will ignore names in that `transportation` layer. 
2. The strings won't be duplicated as the MVT will just add the same strings as tags on more lines. The cost of the extra tag is very minimal.
3. I'll still add the OSM ids for each line in the `transportation` layer because this is useful for our intersection parsing and it definitely helps debugging back to the source maps.

Thanks!
